src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouter.cs

```diff
@@ -0,0 +1,841 @@
+        {
+            _logger = logger;
+
+            _frontendStream = frontendStream;
+            _backendStream = backendStream;
+
+            _cancelRouterTokenSource = new CancellationTokenSource();
+
+            RouterTaskCompleted = new TaskCompletionSource<bool>();
+
+            _backendToFrontendByteTransfer = initBackendToFrontendByteTransfer;
+            _frontendToBackendByteTransfer = initFrontendToBackendByteTransfer;
+
+            Interlocked.Increment(ref s_routerInstanceCount);
+        }
+
+        public void Start()
+        {
+            if (_backendReadFrontendWriteTask != null || _frontendReadBackendWriteTask != null || _disposed)
+                throw new InvalidOperationException();
+
+            _backendReadFrontendWriteTask = BackendReadFrontendWrite(_cancelRouterTokenSource.Token);
+            _frontendReadBackendWriteTask = FrontendReadBackendWrite(_cancelRouterTokenSource.Token);
+        }
+
+        public async void Stop()
+        {
+            if (_disposed)
+                throw new ObjectDisposedException(nameof(ConnectedRouter));
+
+            _cancelRouterTokenSource.Cancel();
+
+            List<Task> runningTasks = new List<Task>();
+
+            if (_backendReadFrontendWriteTask != null)
+                runningTasks.Add(_backendReadFrontendWriteTask);
+
+            if (_frontendReadBackendWriteTask != null)
+                runningTasks.Add(_frontendReadBackendWriteTask);
+
+            await Task.WhenAll(runningTasks.ToArray()).ConfigureAwait(false);
+
+            _backendReadFrontendWriteTask?.Dispose();
+            _frontendReadBackendWriteTask?.Dispose();
+
+            RouterTaskCompleted?.TrySetResult(true);
+
+            _backendReadFrontendWriteTask = null;
+            _frontendReadBackendWriteTask = null;
+        }
+
+        public bool IsRunning {
+            get
+            {
+                if (_backendReadFrontendWriteTask == null || _frontendReadBackendWriteTask == null || _disposed)
+                    return false;
+
+                return !_backendReadFrontendWriteTask.IsCompleted && !_frontendReadBackendWriteTask.IsCompleted;
+            }
+        }
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                Stop();
+
+                _cancelRouterTokenSource.Dispose();
+
+                _backendStream?.Dispose();
+                _frontendStream?.Dispose();
+
+                _disposed = true;
+
+                Interlocked.Decrement(ref s_routerInstanceCount);
+
+                _logger.LogDebug($"Diposed stats: Backend->Frontend {_backendToFrontendByteTransfer} bytes, Frontend->Backend {_frontendToBackendByteTransfer} bytes.");
+                _logger.LogDebug($"Active instances: {s_routerInstanceCount}");
+            }
+        }
+
+        async Task BackendReadFrontendWrite(CancellationToken token)
+        {
+            try
+            {
+                byte[] buffer = new byte[1024];
```

Comment:
If these are happening in quick succession, you could use an [`ArrayPool`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1?view=net-5.0) at the instance level to minimize allocations.

Reply:
This allocation is already reused for the complete router session.